tools/mtmd/clip.cpp

```diff
@@ -3930,15 +3936,146 @@ static std::vector<std::vector<float>> get_2d_sincos_pos_embed(int embed_dim, co
+                    const int ny = patch_size;
+                    const int n = nx * ny;
+                    inp_raw.assign(3 * n, 0.0f);
+
+                    int patch_index = 0;
+                    for (int i = 0; i < image_size_height && patch_index < max_patches; i += patch_size) {
+                        for (int j = 0; j < image_size_width && patch_index < max_patches; j += patch_size) {
+                            for (int pi = 0; pi < patch_size; ++pi) {
+                                for (int pj = 0; pj < patch_size; ++pj) {
+                                    int src_index = ((i + pi) * image_size_width + (j + pj)) * 3;
+                                    int dst_index = nx * pi + patch_index * patch_size + pj;
+                                    inp_raw[dst_index]         = imgs.entries[0]->buf[src_index];
+                                    inp_raw[n + dst_index]     = imgs.entries[0]->buf[src_index + 1];
+                                    inp_raw[2 * n + dst_index] = imgs.entries[0]->buf[src_index + 2];
+                                }
+                            }
+                            patch_index++;
+                        }
+                    }
+                }
+                // prepare position_ids
+                {
+                    // inspired from siglip:
+                    //    -> https://huggingface.co/HuggingFaceM4/siglip-so400m-14-980-flash-attn2-navit
+                    //    -> https://huggingface.co/HuggingFaceM4/siglip-so400m-14-980-flash-attn2-navit/blob/d66538faeba44480d0bfaa42145eef26f9423199/modeling_siglip.py#L316
+                    positions.assign(std::max(pos_h * pos_w, 1024),0);
+                    int bucket_coords_h[1024];
+                    int bucket_coords_w[1024];
+                    for (int i = 0; i < pos_h; i++){
+                        bucket_coords_h[i] = std::floor(70.0*i/pos_h);
+                    }
+                    for (int i = 0; i < pos_w; i++){
+                        bucket_coords_w[i] = std::floor(70.0*i/pos_w);
+                    }
+                    for (int i = 0, id = 0; i < pos_h; i++){
+                        for (int j = 0; j < pos_w; j++){
+                            positions[id++] = bucket_coords_h[i]*70 + bucket_coords_w[j];
+                        }
+                    }
+                }
+                // prepare pos_embed
+                {
+                    // inspired from resampler of Qwen-VL:
+                    //    -> https://huggingface.co/Qwen/Qwen-VL/tree/main
+                    //    -> https://huggingface.co/Qwen/Qwen-VL/blob/0547ed36a86561e2e42fecec8fd0c4f6953e33c4/visual.py#L23
+                    int embed_dim = clip_n_mmproj_embd(ctx);
+
+                    // TODO @ngxson : this is very inefficient, can we do this using ggml_sin and ggml_cos?
+                    auto pos_embed_t = get_2d_sincos_pos_embed(embed_dim, std::make_pair(pos_w, pos_h));
```

Comment:
`get_2d_sincos_pos_embed` is removed on master because it can be done on the cgraph instead. I think it should be also possible to generate this sincos grid on ANE. Can you check?